src/shape.cpp

```diff
@@ -420,6 +420,20 @@ void shape::multi_copy(std::size_t idx, std::size_t* start, const std::size_t* e
     *start = tidx;
 }
 
+bool shape::check_within_bounds(std::vector<std::size_t> multi) const
+{
+    assert(this->lens().size() == multi.size());
+    bool ret = true;
+    for(int i = 0; i < multi.size(); ++i)
+    {
+        if(multi.at(i) > this->lens().at(i))
+        {
+            ret = false;
+        }
+    }
```

Comment:
Use `std::equal` here with `std::less<>{}` for the predicate.

Reply:
For doing a binary `all_of` algorithm, it is named `std::equal` in the standard c++ library. The name `compare` would have been bad since such a comparison is not totally ordered. The name `std::equal` may not be the best name either since its not symmetrical(and the binary predicate is not required to be symmetrical), however that is the name that C++ has chosen for that algorithm(most likely because the common use case of a binary `all_of` is to do equality comparisons). I dont think we should avoid C++ components because they chose a bad or confusing name(ie `vector`, `apply`, etc).

Furthermore, understanding that `std::equal` is a binary `all_of`(C++ developers should be familiar with the standard algorithms), I do find `std::equal` to be clearer here(its also whats suggested [here](https://stackoverflow.com/a/74122451/375343) and its cleaner than using `std::mismatch` or a raw loop).